File: workers/api/src/receipt.ts
export const SLOTS = [
  "who",
  "did",
  "this",
  "when",
  "confirmed_by",
  "if_ok",
  "if_doubt",
  "if_not",
  "status",
] as const;

export type Slot = (typeof SLOTS)[number];
export type Envelope = Record<string, unknown>;
export type ActFields = Record<string, unknown> & { envelope?: Envelope };

export type ReceiptV0 = Record<string, unknown> & {
  id: string;
  receipt_version: "logline.receipt.v0";
  json_canonicalization: "jcs-rfc8785";
  hashes: {
    tuple_hash: string;
    content_hash: string;
    algorithm: "sha256";
  };
};

export type ReceiptV1 = Record<string, unknown> & {
  id: string;
  receipt_version: "logline.receipt.v1";
  json_canonicalization: "jcs-rfc8785";
  envelope: Envelope;
  hashes: {
    tuple_hash: string;
    content_hash: string;
    envelope_hash: string;
    algorithm: "sha256";
  };
};

export type Receipt = ReceiptV0 | ReceiptV1;

const SYSTEM_FIELDS = new Set(["id", "receipt_version", "json_canonicalization", "hashes", "envelope"]);
const FORBIDDEN = new Set(["result", "evidence", "transport"]);

function assertValidString(value: string): void {
  for (let i = 0; i < value.length; i += 1) {
    const code = value.charCodeAt(i);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(i + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) throw new Error("lone surrogate is not valid JCS input");
      i += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      throw new Error("lone surrogate is not valid JCS input");
    }
  }
}

function canonicalize(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "string") {
    assertValidString(value);
    return JSON.stringify(value);
  }
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new Error("non-finite number is not valid JCS input");
    if (Number.isInteger(value) && !Number.isSafeInteger(value)) throw new Error("unsafe integer is not valid JCS input");
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) return `[${value.map(canonicalize).join(",")}]`;
  if (typeof value === "object") {
    const object = value as Record<string, unknown>;
    const keys = Object.keys(object).sort();
    return `{${keys.map((key) => {
      assertValidString(key);
      const item = object[key];
      if (item === undefined || typeof item === "function" || typeof item === "symbol" || typeof item === "bigint") {
        throw new Error(`unsupported JCS value at key ${key}`);
      }
      return `${JSON.stringify(key)}:${canonicalize(item)}`;
    }).join(",")}}`;
  }
  throw new Error(`unsupported JCS value: ${typeof value}`);
}

export function canonicalJson(value: unknown): string {
  return canonicalize(value);
}

export async function sha256Hex(value: string): Promise<string> {
  const data = new TextEncoder().encode(value);
  const digest = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(digest), (byte) => byte.toString(16).padStart(2, "0")).join("");
}

function validateInputFields(fields: ActFields): void {
  for (const key of Object.keys(fields)) {
    if (FORBIDDEN.has(key)) throw new Error(`forbidden top-level field: ${key}`);
  }
  for (const slot of SLOTS) {
    if (slot in fields && typeof fields[slot] !== "string") {
      throw new Error(`receipt slot ${slot} must be a string`);
    }
  }
  if ("envelope" in fields && (fields.envelope === null || typeof fields.envelope !== "object" || Array.isArray(fields.envelope))) {
    throw new Error("envelope must be a JSON object");
  }
}

function contentMaterial(input: ActFields): Record<string, unknown> {
  const content: Record<string, unknown> = {};
  for (const slot of SLOTS) content[slot] = typeof input[slot] === "string" ? input[slot] : "";
  for (const [key, value] of Object.entries(input)) {
    if (!SLOTS.includes(key as Slot) && !SYSTEM_FIELDS.has(key) && !FORBIDDEN.has(key)) content[key] = value;
  }
  return content;
}

/**
 * Mint the canonical v1 universal unit.
 *
 * content_hash  = H(JCS(LogLine 9 fields + AUX))
 * envelope_hash = H(JCS(envelope))
 * tuple_hash    = H(content_hash + envelope_hash)
 *
 * Receipt metadata is intentionally outside content identity. `envelope` is persisted
 * with the receipt so the contextual half of tuple identity can always be replayed.
 */
export async function mintReceipt(input: ActFields): Promise<ReceiptV1> {
  validateInputFields(input);
  const content = contentMaterial(input);
  const envelope: Envelope = input.envelope ? { ...input.envelope } : {};
  const contentHash = await sha256Hex(canonicalJson(content));
  const envelopeHash = await sha256Hex(canonicalJson(envelope));
  const tupleHash = await sha256Hex(`${contentHash}${envelopeHash}`);
  return {
    ...content,
    envelope,
    receipt_version: "logline.receipt.v1",
    json_canonicalization: "jcs-rfc8785",
    hashes: {
      tuple_hash: tupleHash,
      content_hash: contentHash,
      envelope_hash: envelopeHash,
      algorithm: "sha256",
    },
    id: contentHash,
  } as ReceiptV1;
}

/** Historical v0 minting retained only for parity/verification of existing ledger data. */
export async function mintReceiptV0(input: ActFields): Promise<ReceiptV0> {
  validateInputFields(input);
  const receipt: Record<string, unknown> = {};
  for (const slot of SLOTS) receipt[slot] = typeof input[slot] === "string" ? input[slot] : "";
  receipt.receipt_version = "logline.receipt.v0";
  receipt.json_canonicalization = "jcs-rfc8785";

  for (const [key, value] of Object.entries(input)) {
    if (!SLOTS.includes(key as Slot) && !SYSTEM_FIELDS.has(key) && !FORBIDDEN.has(key)) receipt[key] = value;
  }

  const tupleMaterial = Object.fromEntries(SLOTS.map((slot) => [slot, receipt[slot]]));
  const legacyContentMaterial = Object.fromEntries(Object.entries(receipt).filter(([key]) => key !== "id" && key !== "hashes"));
  const tupleHash = await sha256Hex(canonicalJson(tupleMaterial));
  const contentHash = await sha256Hex(canonicalJson(legacyContentMaterial));
  receipt.hashes = { tuple_hash: tupleHash, content_hash: contentHash, algorithm: "sha256" };
  receipt.id = contentHash;
  return receipt as ReceiptV0;
}

File: lab/receipt.py
"""Canonical LogLine receipt minting and verification.

v1 implements the product v1.2 identity split:

    content_hash  = H(JCS(LogLine 9 fields + AUX))
    envelope_hash = H(JCS(envelope))
    tuple_hash    = H(content_hash + envelope_hash)

Receipt metadata is not content identity. The envelope is persisted in the universal
unit so process/custody context can be replayed. Historical ``logline.receipt.v0``
verification remains supported byte-for-byte; new ``mint`` calls produce v1.
"""
from __future__ import annotations

import hashlib
from collections.abc import Mapping
from typing import Any

from ._vendor import rfc8785
from .errors import ReceiptError

SLOTS = ("who", "did", "this", "when", "confirmed_by", "if_ok", "if_doubt", "if_not", "status")
SYSTEM_FIELDS = {"id", "receipt_version", "json_canonicalization", "hashes", "envelope"}
FORBIDDEN = {"result", "evidence", "transport"}
HASH_FIELDS_V0 = {"tuple_hash", "content_hash", "algorithm"}
HASH_FIELDS_V1 = {"tuple_hash", "content_hash", "envelope_hash", "algorithm"}
# Compatibility export used by historical hardening tests.
HASH_FIELDS = HASH_FIELDS_V1
RECEIPT_VERSION_V0 = "logline.receipt.v0"
RECEIPT_VERSION_V1 = "logline.receipt.v1"
RECEIPT_VERSION = RECEIPT_VERSION_V1
CANONICALIZATION = "jcs-rfc8785"
HASH_ALGORITHM = "sha256"


def canonical_json(value: Any) -> str:
    """Return RFC 8785 (JCS) canonical JSON for identity hash material."""
    try:
        return rfc8785.dumps(value).decode("utf-8")
    except rfc8785.CanonicalizationError as exc:
        raise ReceiptError(f"value is not canonicalizable under RFC 8785: {exc}") from exc


def sha256_text(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def _validate_input_fields(fields: Mapping[str, Any]) -> None:
    forbidden = FORBIDDEN.intersection(fields)
    if forbidden:
        raise ReceiptError(f"forbidden top-level field(s): {', '.join(sorted(forbidden))}")
    for slot in SLOTS:
        if slot in fields and not isinstance(fields[slot], str):
            raise ReceiptError(f"receipt slot {slot!r} must be a string")
    if "envelope" in fields and not isinstance(fields["envelope"], Mapping):
        raise ReceiptError("envelope must be a JSON object")


def _content_material(fields: Mapping[str, Any]) -> dict[str, Any]:
    content: dict[str, Any] = {slot: fields.get(slot, "") for slot in SLOTS}
    for key, value in fields.items():
        if key not in SLOTS and key not in SYSTEM_FIELDS and key not in FORBIDDEN:
            content[key] = value
    return content


def mint(fields: Mapping[str, Any], *, envelope: Mapping[str, Any] | None = None) -> dict[str, Any]:
    """Mint the canonical v1 universal unit from LogLine+AUX plus envelope."""
    _validate_input_fields(fields)
    content = _content_material(fields)
    source_envelope = envelope if envelope is not None else fields.get("envelope", {})
    if not isinstance(source_envelope, Mapping):
        raise ReceiptError("envelope must be a JSON object")
    envelope_value = dict(source_envelope)

    content_hash = sha256_text(canonical_json(content))
    envelope_hash = sha256_text(canonical_json(envelope_value))
    tuple_hash = sha256_text(content_hash + envelope_hash)

    receipt: dict[str, Any] = {
        **content,
        "envelope": envelope_value,
        "receipt_version": RECEIPT_VERSION_V1,
        "json_canonicalization": CANONICALIZATION,
        "hashes": {
            "tuple_hash": tuple_hash,
            "content_hash": content_hash,
            "envelope_hash": envelope_hash,
            "algorithm": HASH_ALGORITHM,
        },
        "id": content_hash,
    }
    return receipt


def mint_v0(fields: Mapping[str, Any]) -> dict[str, Any]:
    """Mint historical receipt v0 exactly; retained for parity fixtures/migrations."""
    _validate_input_fields(fields)
    receipt: dict[str, Any] = {slot: fields.get(slot, "") for slot in SLOTS}
    receipt["receipt_version"] = RECEIPT_VERSION_V0
    receipt["json_canonicalization"] = CANONICALIZATION

    for key, value in fields.items():
        if key not in SLOTS and key not in SYSTEM_FIELDS and key not in FORBIDDEN:
            receipt[key] = value

    tuple_material = {slot: receipt[slot] for slot in SLOTS}
    legacy_content_material = {key: value for key, value in receipt.items() if key not in {"id", "hashes"}}
    tuple_hash = sha256_text(canonical_json(tuple_material))
    content_hash = sha256_text(canonical_json(legacy_content_material))
    receipt["hashes"] = {"tuple_hash": tuple_hash, "content_hash": content_hash, "algorithm": HASH_ALGORITHM}
    receipt["id"] = content_hash
    return receipt


def verify(receipt: Mapping[str, Any]) -> tuple[bool, str]:
    """Verify either supported receipt version without throwing."""
    try:
        verify_or_raise(receipt)
    except ReceiptError as exc:
        return False, str(exc)
    return True, "ok"


def _require_common(receipt: Mapping[str, Any]) -> Mapping[str, Any]:
    missing = [field for field in (*SLOTS, "receipt_version", "json_canonicalization", "hashes", "id") if field not in receipt]
    if missing:
        raise ReceiptError(f"missing required field(s): {', '.join(missing)}")
    _validate_input_fields(receipt)
    if receipt["json_canonicalization"] != CANONICALIZATION:
        raise ReceiptError("unsupported json_canonicalization")
    hashes = receipt["hashes"]
    if not isinstance(hashes, Mapping) or hashes.get("algorithm") != HASH_ALGORITHM:
        raise ReceiptError("hashes.algorithm must be sha256")
    return hashes


def _verify_v0(receipt: Mapping[str, Any], hashes: Mapping[str, Any]) -> None:
    extra_hash_fields = set(hashes) - HASH_FIELDS_V0
    if extra_hash_fields:
        raise ReceiptError(f"hashes object has forbidden field(s): {', '.join(sorted(extra_hash_fields))}")
    expected = mint_v0({key: value for key, value in receipt.items() if key not in {"id", "hashes"}})
    if receipt["id"] != expected["id"]:
        raise ReceiptError("id/content_hash mismatch")
    if hashes.get("tuple_hash") != expected["hashes"]["tuple_hash"]:
        raise ReceiptError("tuple_hash mismatch")
    if hashes.get("content_hash") != expected["hashes"]["content_hash"]:
        raise ReceiptError("content_hash mismatch")


def _verify_v1(receipt: Mapping[str, Any], hashes: Mapping[str, Any]) -> None:
    if "envelope" not in receipt or not isinstance(receipt["envelope"], Mapping):
        raise ReceiptError("receipt v1 requires envelope object")
    extra_hash_fields = set(hashes) - HASH_FIELDS_V1
    missing_hash_fields = HASH_FIELDS_V1 - set(hashes)
    if extra_hash_fields:
        raise ReceiptError(f"hashes object has forbidden field(s): {', '.join(sorted(extra_hash_fields))}")
    if missing_hash_fields:
        raise ReceiptError(f"hashes object missing field(s): {', '.join(sorted(missing_hash_fields))}")

    expected = mint({key: value for key, value in receipt.items() if key not in {"id", "hashes"}})
    if receipt["id"] != expected["id"]:
        raise ReceiptError("id/content_hash mismatch")
    for field in ("content_hash", "envelope_hash", "tuple_hash"):
        if hashes.get(field) != expected["hashes"][field]:
            raise ReceiptError(f"{field} mismatch")


def verify_or_raise(receipt: Mapping[str, Any]) -> None:
    hashes = _require_common(receipt)
    version = receipt["receipt_version"]
    if version == RECEIPT_VERSION_V0:
        _verify_v0(receipt, hashes)
        return
    if version == RECEIPT_VERSION_V1:
        _verify_v1(receipt, hashes)
        return
    raise ReceiptError("unsupported receipt_version")
